### PR TITLE
"Improve language in 'Constant Expressions'.

### LINKED ISSUE #497: Language in constant evaluation section is unclear
This text from [**Constant Evaluation**: Constant Expressions][ce-ce] is a bit murky:

[ce-ce]: https://doc.rust-lang.org/reference/const_eval.html#constant-expressions

> Certain types of expressions can be evaluated at compile time. These are called constant expressions and are required in const contexts. In other places, such as in [`let` statements], constant expressions may be evaluated at compile time. If errors, such as out of bounds [array indexing] or [overflow] occurs, then it is a compiler error if the value must be evaluated at compile time, otherwise it is just a warning, but the code will most likely panic when run.
> 
> [`let` statements]: 
https://doc.rust-lang.org/reference/statements.html#let-statements
> [array indexing]: 
https://doc.rust-lang.org/reference/expressions/array-expr.html#array-and-slice-indexing-expressions
> [overflow]: 
https://doc.rust-lang.org/reference/expressions/operator-expr.html#overflow

Suggested rephrasing:

> Certain types of expressions, called constant expressions, can be evaluated at compile time. In [const contexts], these are the only allowed expressions, and are always evaluated at compile time. In other places, such as in [`let` statements], constant expressions are allowed but not required, and *may* be evaluated at compile time. Errors such as out of bounds [array indexing] or [overflow] are compiler errors if the value must be evaluated at compile time (i.e. in const contexts); otherwise they are just warnings, but the code will most likely panic when run.
>
> [const contexts]: https://doc.rust-lang.org/reference/const_eval.html#const-context
> [`let` statements]: 
https://doc.rust-lang.org/reference/statements.html#let-statements
> [array indexing]: 
https://doc.rust-lang.org/reference/expressions/array-expr.html#array-and-slice-indexing-expressions
> [overflow]: 
https://doc.rust-lang.org/reference/expressions/operator-expr.html#overflow

Happy to submit a PR to that effect or similar!

### PR DESCRIPTION
**Pull request by chriskrycho**
Fixes #497.

### CODE REVIEW

File: src/const_eval.md [hunk 1]
```diff
@@ -6,13 +6,13 @@ can be evaluated at compile-time.
...
 ## Constant expressions
 
-Certain types of expressions can be evaluated at compile time. These are called
-_constant expressions_ and are required in const contexts. In
-other places, such as in [`let` statements](statements.html#let-statements),
-constant expressions may be evaluated at compile time. If errors, such as out
-of bounds [array indexing] or [overflow] occurs,
-then it is a compiler error if the value must be evaluated at compile time,
-otherwise it is just a warning, but the code will most likely panic when run.
+Certain types of expressions, called constant expressions, can be evaluated at
```
Comment (Centril) [inline src/const_eval.md]:
```suggestion
Certain forms of expressions, called constant expressions, can be evaluated at
```
to not confuse with expression typing...

File: src/const_eval.md [hunk 2]
```diff
@@ -6,13 +6,13 @@ can be evaluated at compile-time.
...
+compile time. In [const contexts](#const-context), these are the only allowed
+expressions, and are always evaluated at compile time. In other places, such as
+in [`let` statements](statements.html#let-statements), constant expressions are
```
Comment (Centril) [inline src/const_eval.md]:
```suggestion
[`let` statements](statements.html#let-statements), constant expressions are
```

File: src/const_eval.md [hunk 3]
```diff
@@ -6,13 +6,13 @@ can be evaluated at compile-time.
...
+Errors such as out of bounds [array indexing] or [overflow] are compiler errors
```
Comment (Centril) [inline src/const_eval.md]: I think:
> allowed but not required, and may be evaluated at compile time.

accidentally fell off?

File: src/const_eval.md [hunk 4]
```diff
@@ -6,13 +6,13 @@ can be evaluated at compile-time.
...
+if the value must be evaluated at compile time (i.e. in const contexts);
+otherwise they are just warnings, but the code will most likely panic when run.
```
Comment (Centril) [inline src/const_eval.md]:
```suggestion
Otherwise, these behaviors are warnings, but will likely panic when run.
```

File: src/const_eval.md [hunk 5]
```diff
@@ -6,13 +6,13 @@ can be evaluated at compile-time.
...
-constant expressions may be evaluated at compile time. If errors, such as out
-of bounds [array indexing] or [overflow] occurs,
-then it is a compiler error if the value must be evaluated at compile time,
-otherwise it is just a warning, but the code will most likely panic when run.
+Certain types of expressions, called constant expressions, can be evaluated at
+compile time. In [const contexts](#const-context), these are the only allowed
+expressions, and are always evaluated at compile time. In other places, such as
+in [`let` statements](statements.html#let-statements), constant expressions are
+Errors such as out of bounds [array indexing] or [overflow] are compiler errors
+if the value must be evaluated at compile time (i.e. in const contexts);
```
Comment (Centril) [inline src/const_eval.md]:
```suggestion
if the value must be evaluated at compile time (i.e. in const contexts).
```

File: src/const_eval.md (hunk 3 above)
Comment (Centril) [reply src/const_eval.md]: Also, given that constant expressions are a strict subset of all expressions, is "allowed but not required" not implied?

Comment (Centril) [inline src/const_eval.md]:
```suggestion
Behaviors such as out of bounds [array indexing] or [overflow] are compiler errors
```

Comment (chriskrycho) [reply src/const_eval.md]: Ah, good catch on the missing bit; I'll push a change with that and also clarifying the other language.

Comment (chriskrycho) [reply src/const_eval.md]: And, there we go.

**Comment by chriskrycho** [comment]
Yeah, I realized I was more motivated to push this than to clean the dishes! 😆

**Comment by Centril** [comment]
Haha; I sympathize with not cleaning the dishes ;)

This turned out great; thanks!

**Comment by chriskrycho** [comment]
My pleasure!

**Pull request merged**